### ISSUE TITLE
[DLPack] Strides are corrupted after `from_dlpack` / `to_dlpack`

### ISSUE THREAD

**Issue by masahi** [pytorch/pytorch#99803]
### 🐛 Describe the bug

Strides of a tensor is corrupted after using `from_dlpack` / `to_dlpack`:

```
import torch
from torch.utils.dlpack import to_dlpack, from_dlpack

a = torch.randint(0, 100, (1, 77), dtype=torch.int32)
b = from_dlpack(to_dlpack(a))

print(a.stride())
print(b.stride())
```

It prints:
```
(77, 1)
(1, 1)
```

According to https://discuss.tvm.apache.org/t/a-pytorch-and-tvm-version-incompatibility-problem-while-using-from-dlpack-and-to-dlpack/14771, PT 1.13 has the same problem while 1.12 does not. I confirmed that, with PT 1.12 I get correct strides back.  

### Versions

```
Collecting environment information...
PyTorch version: 2.0.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.0.1 (arm64)
GCC version: Could not collect
Clang version: 15.0.5
CMake version: version 3.25.0
Libc version: N/A

Python version: 3.10.6 | packaged by conda-forge | (main, Aug 22 2022, 20:38:29) [Clang 13.0.1 ] (64-bit runtime)
Python platform: macOS-13.0.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1 Pro

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.23.5
[pip3] pytorch-lightning==1.8.3.post1
[pip3] torch==2.0.0
[pip3] torch-mlir==20221203.676
[pip3] torchaudio==2.0.1
[pip3] torchmetrics==0.10.3
[pip3] torchvision==0.15.1
[conda] numpy                     1.23.5                   pypi_0    pypi
[conda] pytorch-lightning         1.8.3.post1              pypi_0    pypi
[conda] torch                     2.0.0                    pypi_0    pypi
[conda] torch-mlir                20221203.676             pypi_0    pypi
[conda] torchaudio                2.0.1                    pypi_0    pypi
[conda] torchmetrics              0.10.3                   pypi_0    pypi
[conda] torchvision               0.15.1                   pypi_0    pypi
```

**Comment by masahi** [comment]
It seems it is due to https://github.com/pytorch/pytorch/pull/83158

**Comment by mattip** [comment]
Since #83158 is needed to fix #82610, I don't think reverting it is the path forward. The stride of a tensor where the shape is 0 or 1 is not well defined, it could be any value. PR 83158 forces the stride to 1. This is intentional, and done to match the behaviour of Numpy `memoryview(a)`. Is the problem a cosmetic one or is there code that depends on round-tripping the strides? cc @dalcinl who raised the original issue.

**Comment by masahi** [comment]
I'll read discussions in #82610 to understand the context, but for a row-major tensor of shape (1, 77), the stride should be clearly (77, 1) and (1, 1) makes no sense. 

> Is the problem a cosmetic one or is there code that depends on round-tripping the strides? cc @dalcinl who raised the original issue

External libraries like https://github.com/apache/tvm needs the correct stride. Currently passing PyTorch tensors of certain shapes to TVM doesn't work because of this. See for example https://discuss.tvm.apache.org/t/a-pytorch-and-tvm-version-incompatibility-problem-while-using-from-dlpack-and-to-dlpack/14771

**Comment by masahi** [comment]
That said, I'll discuss with TVM people if the size 1 / stride 1 case should be handled by the TVM runtime, cc @tqchen

**Comment by masahi** [comment]
Closing since I understood that this is not a bug.

**Comment by dalcinl** [comment]
@masahi Just my two cents: PyTorch should not trust its consumers, and then implement strides as close as NumPy as possible, in all cases, following the same normalization rules (although I understand this may be a lot of work). TVM should not trust its providers (like I did in mpi4py) and ignore/handle properly  the case `size={0,1}` in any of its dimension.
